src/main/java/frc/robot/RobotContainer.java

```diff
@@ -105,7 +116,30 @@ private void configureDefaultCommands() {
                         () -> true));
     }
 
-    private void configureButtonBindings() {}
+    private Command updateScoreState() {
+        return (currentState.calculateTargets().andThen(currentState.prepareSubsystems()))
+                .repeatedly();
+    }
+
+    private void configureButtonBindings() {
+        xboxController
+                .a()
+                .onTrue(
+                        Commands.runOnce(() -> currentState = shootState)
+                                .andThen(updateScoreState()));
+        xboxController
+                .b()
+                .onTrue(
+                        Commands.runOnce(() -> currentState = ampState)
+                                .andThen(updateScoreState()));
+        xboxController
+                .x()
+                .onTrue(
+                        Commands.runOnce(() -> currentState = climbState)
+                                .andThen(updateScoreState()));
```

Comment:
403c7b7

Reply:
Amp/Climb
I'm not sure why you overloaded `updateScoreState()` in shoot state, the previous structure was fine -- `updateInBounds()` is already called in `calculateTargets()` which is fine. I only suggested adding the current state update to `updateScoreState()` in robot container.

Reply:
I don't understand what do you want

Reply:
what is current state updating?

Reply:
b1eda81d2b8e0cbf1e3aba096a727804256bb4b1

Reply:
Perhaps there is some misunderstanding. It was a bit of a lot of code to write here so I pushed my suggestion to my fork. Changes I made:
1. Moved new state setting into `updateScoreState()`.
2. Changed this command to a deferred command. I think this was needed before, too, because `Commands.repeatingSequence(currentState.calculateTargets(), currentState.prepareSubsystems())` was evaluated once based on the initial state so it wouldn't have updated after state changes.
3. Added a parameter to `updateScoreState()` that receives the requirement set. Note that this could have been done with a constant set of _all_ scoring subsystems but a parameter enabled this to be more granular.
4. I've also converted the scoring command to a deferred command because it fits here better than a proxy command.
The relevant diff is [here](https://github.com/Galaxia5987/robot-2024/compare/dca4391...97128af), you can see the full file [here](https://github.com/katzuv/robot-2024/blob/97128af2ae8b32af6f2b4bfbe21b912124215125/src/main/java/frc/robot/RobotContainer.java).

<details>
  <summary>Here is the full diff directly, too:</summary>

```diff
diff --git a/src/main/java/frc/robot/RobotContainer.java b/src/main/java/frc/robot/RobotContainer.java                                         
index c1f8493..54b1746 100644
--- a/src/main/java/frc/robot/RobotContainer.java
+++ b/src/main/java/frc/robot/RobotContainer.java
@@ -2,6 +2,7 @@ package frc.robot;

 import edu.wpi.first.wpilibj2.command.Command;
 import edu.wpi.first.wpilibj2.command.Commands;
+import edu.wpi.first.wpilibj2.command.Subsystem;
 import edu.wpi.first.wpilibj2.command.button.CommandXboxController;
 import frc.robot.scoreStates.AmpState;
 import frc.robot.scoreStates.ClimbState;
@@ -30,6 +31,7 @@ import frc.robot.subsystems.shooter.ShooterIO;
 import frc.robot.subsystems.shooter.ShooterIOReal;
 import frc.robot.subsystems.shooter.ShooterIOSim;
 import frc.robot.subsystems.swerve.SwerveDrive;
+import java.util.Set;

 public class RobotContainer {
     private static RobotContainer INSTANCE = null;
@@ -115,29 +117,32 @@ public class RobotContainer {
                         () -> true));
     }

-    private Command updateScoreState() {
-        return Commands.repeatingSequence(
-                currentState.calculateTargets(), currentState.prepareSubsystems());
+    private Command updateScoreState(ScoreState newState, Set<Subsystem> scoringRequirements) {
+        return Commands.defer(
+                () ->
+                        Commands.runOnce(() -> currentState = newState)
+                                .andThen(
+                                        Commands.repeatingSequence(
+                                                currentState.calculateTargets(),
+                                                currentState.prepareSubsystems())),
+                scoringRequirements);
     }

     private void configureButtonBindings() {
         xboxController
                 .a()
-                .onTrue(
-                        Commands.runOnce(() -> currentState = shootState)
-                                .andThen(updateScoreState()));
+                .onTrue(updateScoreState(shootState, Set.of(swerveDrive, conveyor, hood, shooter)));
+
+        Set<Subsystem> ampClimbRequirements = Set.of(swerveDrive, elevator, gripper);
+        xboxController.b().onTrue(updateScoreState(ampState, ampClimbRequirements));
+        xboxController.x().onTrue(updateScoreState(climbState, ampClimbRequirements));
+
         xboxController
                 .b()
                 .onTrue(
-                        Commands.runOnce(() -> currentState = ampState)
-                                .andThen(updateScoreState()));
-        xboxController
-                .x()
-                .onTrue(
-                        Commands.runOnce(() -> currentState = climbState)
-                                .andThen(updateScoreState()));
-
-        xboxController.rightTrigger(0.1).onTrue(Commands.deferredProxy(() -> currentState.score()));
+                        Commands.defer(
+                                () -> currentState.score(),
+                                Set.of(swerveDrive, conveyor, hood, shooter, elevator, gripper)));
     }
```
</details>

Reply:
e378d24
I also fixed the swerve being null & shootingCSV
57ca21b

Reply:
Awesome! See what I added in the edit, I took the state update out of the defer

Reply:
I saw, earlier i just didn't understand what did you mean

Reply:
Look how it's in my fork
https://github.com/katzuv/robot-2024/blob/05c234e5fc3a8a31b91973300801c76048f01fca/src/main/java/frc/robot/RobotContainer.java#L120-L125

Reply:
In the code the current state will not update no?